src/web/extractors.rs

```diff
@@ -358,6 +411,23 @@ pub struct CollectionPostRequest {
     pub bsos: BsoBodies,
 }
 
+impl CollectionPostRequest {
+    /// Return bsos that failed to parse as appropriate failed hashmap
+    pub fn failed_bsos(&self) -> HashMap<String, String> {
+        // Leave room for any new errors when attempting to save them
+        let mut result = HashMap::with_capacity(self.bsos.invalid.len() + self.bsos.valid.len());
+        for raw in self.bsos.invalid.iter() {
+            if let Some(id) = raw
+                .get("id")
+                .and_then(|v| v.as_str().map(|s| s.to_string()))
```

Comment:
moving the final to_string() call into the insert line below is a little clearer I think